pageserver/src/deletion_queue.rs

```diff
@@ -0,0 +1,1310 @@
+    });
+
+    struct TestSetup {
+        harness: TenantHarness,
+        remote_fs_dir: PathBuf,
+        storage: GenericRemoteStorage,
+        mock_control_plane: MockControlPlane,
+        deletion_queue: DeletionQueue,
+        worker_join: JoinHandle<()>,
+    }
+
+    impl TestSetup {
+        /// Simulate a pageserver restart by destroying and recreating the deletion queue
+        async fn restart(&mut self) {
+            let (deletion_queue, workers) = DeletionQueue::new(
+                Some(self.storage.clone()),
+                Some(self.mock_control_plane.clone()),
+                self.harness.conf,
+            );
+
+            tracing::debug!("Spawning worker for new queue queue");
+            let worker_join = workers
+                .unwrap()
+                .spawn_with(&tokio::runtime::Handle::current());
+
+            let old_worker_join = std::mem::replace(&mut self.worker_join, worker_join);
+            let old_deletion_queue = std::mem::replace(&mut self.deletion_queue, deletion_queue);
+
+            tracing::debug!("Joining worker from previous queue");
+            old_deletion_queue.cancel.cancel();
+            old_worker_join
+                .await
+                .expect("Failed to join workers for previous deletion queue");
+        }
+
+        fn set_latest_generation(&self, gen: Generation) {
+            let tenant_id = self.harness.tenant_id;
+            self.mock_control_plane
+                .latest_generation
+                .lock()
+                .unwrap()
+                .insert(tenant_id, gen);
+        }
+
+        /// Returns remote layer file name, suitable for use in assert_remote_files
+        fn write_remote_layer(
+            &self,
+            file_name: LayerFileName,
+            gen: Generation,
+        ) -> anyhow::Result<String> {
+            let tenant_id = self.harness.tenant_id;
+            let relative_remote_path = remote_timeline_path(&tenant_id, &TIMELINE_ID);
+            let remote_timeline_path = self.remote_fs_dir.join(relative_remote_path.get_path());
+            std::fs::create_dir_all(&remote_timeline_path)?;
+            let remote_layer_file_name = format!("{}{}", file_name, gen.get_suffix());
+
+            let content: Vec<u8> = format!("placeholder contents of {file_name}").into();
+
+            std::fs::write(
+                remote_timeline_path.join(remote_layer_file_name.clone()),
+                content,
+            )?;
+
+            Ok(remote_layer_file_name)
+        }
+    }
+
+    #[derive(Debug, Clone)]
+    struct MockControlPlane {
+        pub latest_generation: std::sync::Arc<std::sync::Mutex<HashMap<TenantId, Generation>>>,
+    }
+
+    impl MockControlPlane {
+        fn new() -> Self {
+            Self {
+                latest_generation: Arc::new(std::sync::Mutex::new(HashMap::new())),
+            }
+        }
+    }
+
+    unsafe impl Send for MockControlPlane {}
+    unsafe impl Sync for MockControlPlane {}
```

Comment:
Why are these needed?

Reply:
Perhaps usually anything dyn is behind an `Arc<dyn ControlPlaneGenerationsApi>` but this is used as is, wonder if that could be it? Then the `self` in the trait impl would be `Send` (which will also need to be `Sync` because `&self` is needed to be Send).

Reply:
```suggestion
```

Just checked, these are no longer needed. The `Arc<dyn Foo>` is something to remember in future though I cannot really see what was the reason you added these in the first place.